### PR TITLE
P20-1183: Enable Global Edition region switching in Pegasus

### PR DESCRIPTION
**Pull request by artem-vavilov**
## Summary

| Farsi region  | en-US |
| ------------- | ------------- |
| Header |  |
| Footer |  |

| Farsi region  | fa-IR |
| ------------- | ------------- |
| Header |  |
| Footer |  |

## Links
- JIRA task: [P20-1183](https://codedotorg.atlassian.net/browse/P20-1183)
- Related PR: https://github.com/code-dot-org/code-dot-org/pull/61397

### CODE REVIEW

File: config/global_editions/root.yml
Comment (wilkie) [inline config/global_editions/root.yml]: I think this should be `locales`?

File: shared/middleware/varnish_environment.rb
Comment (wilkie) [inline shared/middleware/varnish_environment.rb]: It feels confusing to have both a helper named `cookie_locale` and one it uses named `locale_cookie` ... maybe we can have a different naming pattern. Perhaps `locale_from_cookie` and `locale_from_varnish`? or `cookie_locale_value` and `varnish_locale_value`... not sure.

I think they are only ever really used in the middleware config, so hopefully they are not a pain to name?

File: lib/cdo/rack/global_edition.rb
Comment (wilkie) [inline lib/cdo/rack/global_edition.rb:123]: where is this used?

**Review by wilkie — APPROVED** [review]
This is a good refactor of the localization configuration and language setting paths in the codebase between Pegasus and Dashboard. I especially like the clean-up of the helpers deciding language direction and names to unify them.

This looks good, and I know it to be working from our Adhoc testing.

My little nitpicks on the names of the helpers are just to consider some names such that `cookie_locale` and `locale_cookie` aren't confusing in the future, but I'm more or less OK with that just being a by-product of naming schemes colliding awkwardly.

We can expect that staging will sputter a few errors as it git pulls and builds with changes to the middleware and CDO namespaces.

File: lib/cdo/rack/global_edition.rb
Comment (artem-vavilov) [reply lib/cdo/rack/global_edition.rb:123]: I missed this part after many merge conflict fixes: https://github.com/code-dot-org/code-dot-org/pull/62061/commits/34c8352af109cf9d779200330f722d4a8638fdb7

File: shared/middleware/varnish_environment.rb
Comment (artem-vavilov) [reply shared/middleware/varnish_environment.rb]: https://github.com/code-dot-org/code-dot-org/pull/62061/commits/0cecaf1bd36313442200e8a1a20d0f1c3eb40228

**Review by wilkie — APPROVED** [review]

**Comment by wilkie** [comment]
Per external conversation in Slack, we have decided to:
* add a DCDO flag (`global_edition_enabled`) to guard the use of the middleware (in the `def call` for the middleware around the same point as the host check)
* ensure that exceptions simply log and continue the call chain, to avoid site outages since this middleware is only redirecting and can fail gracefully

**Review by wilkie — CHANGES_REQUESTED** [review]
Per external conversation in Slack, we have decided to:
* add a DCDO flag (`global_edition_enabled`) to guard the use of the middleware (in the `def call` for the middleware around the same point as the host check)
* ensure that exceptions simply log and continue the call chain, to avoid site outages since this middleware is only redirecting and can fail gracefully

**Comment by wilkie** [comment]
Created an adhoc from this branch. Upon loading without the flag:

image

Then enabled the flag:

```
DCDO.set('global_edition_enabled', true)
```

When results in the dropdown now existing:

image

The region switching logic is allowed and I see the appropriate page.

Trying to navigate back to the root is disallowed.

Disengaging the flag:

```
DCDO.set('global_edition_enabled', false)
```

Still forcibly goes back to `/global/fa` when I try to go to the `/` root page.

This is because Chrome is still caching the `302`. Disabling the cache now shows the correct behavior. We should not locally cache this `302` redirect.

I think we need to respond with an injected `Cache-Control` value of `no-cache, no-store` to ensure the pages redirect properly when either the flag is disabled or their region changes elsewhere.

Still, the flag more or less is doing what it is meant to do and would prevent new users from interacting with the redirects.

**Comment by wilkie** [comment]
Imposing a raised exception in the path of the middleware does gracefully allow it to continue serving the site:

image

The dropdown still shows up, and selecting Farsi shows evidence that it tries to load the right page but the fallback prevents the redirect but it still loads a reasonable page as a result. Seems good!

**Comment by wilkie** [comment]
Testing out adding `response.do_not_cache!` to paths that perform a redirect. This sets `no-cache must-reevaluate` with an expiry, but I'm wondering if I need a `no-store`. I suppose I might not.

**Comment by wilkie** [comment]
Adding `no-cache` via `do_not_cache!` and repeating my prior experiment:

image

I notice the redirect cache notice. I then disable the experiment:

```
DCDO.set('global_edition_enabled', false)
```

Seems to work:

image

Using this commit: d27cd1b50666164ef94a8ff07b24c68ce248b360

**Review by wilkie — APPROVED** [review]

**Review by wilkie — APPROVED** [review]

**Comment by wilkie** [comment]
I've been testing this on the adhoc and it feels good. Feel free to also test what you'd like on the adhoc with all of the changes combined.

I say, merge this when the tests reliably pass and that will give me a chance to rebase all the other branches to merge them tomorrow.

**Pull request merged**
